<!DOCTYPE html>
<html lang="zh-CN">
<head>
  <meta charset="UTF-8">
<meta name="viewport" content="width=device-width">
<meta name="theme-color" content="#222">
<meta name="generator" content="Hexo 5.0.2">


  <link rel="apple-touch-icon" sizes="180x180" href="/images/apple-touch-icon-next.png">
  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon-32x32-next.png">
  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon-16x16-next.png">
  <link rel="mask-icon" href="/images/logo.svg" color="#222">

<link rel="stylesheet" href="/css/main.css">



<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free@5.15.3/css/all.min.css" integrity="sha256-2H3fkXt6FEmrReK448mDVGKb3WW2ZZw35gI7vqHOE4Y=" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/animate.css@3.1.1/animate.min.css" integrity="sha256-PR7ttpcvz8qrF57fur/yAx1qXMFJeJFiA6pSzWi0OIE=" crossorigin="anonymous">

<script class="next-config" data-name="main" type="application/json">{"hostname":"bubbleboy11.github.io","root":"/","images":"/images","scheme":"Muse","version":"8.6.1","exturl":false,"sidebar":{"position":"left","display":"post","padding":18,"offset":12},"copycode":true,"bookmark":{"enable":false,"color":"#222","save":"auto"},"fancybox":false,"mediumzoom":false,"lazyload":false,"pangu":false,"comments":{"style":"tabs","active":null,"storage":true,"lazyload":false,"nav":null},"motion":{"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"fadeInDown","post_body":"fadeInDown","coll_header":"fadeInLeft","sidebar":"fadeInUp"}},"prism":false,"i18n":{"placeholder":"搜索...","empty":"没有找到任何搜索结果：${query}","hits_time":"找到 ${hits} 个搜索结果（用时 ${time} 毫秒）","hits":"找到 ${hits} 个搜索结果"}}</script><script src="/js/config.js"></script>
<meta name="description" content="错误和异常是两个不同的概念，非常容易混淆。很多程序员习惯将一切非正常情况都看做错误，一切皆错误或一切皆异常的情况，而不区分错误和异常，即使程序中可能有异常抛出，也将异常及时捕获并转换成错误。从表面上看，一切皆错误的思路更简单，而异常的引入仅仅增加了额外的复杂度。但事实并非如此。众所周知，Golang遵循“少即是多”的设计哲学，追求简洁优雅，就是说如果异常价值不大，就不会将异常加入到语言特性中。 错">
<meta property="og:type" content="article">
<meta property="og:title" content="go错误和异常panic和recover">
<meta property="og:url" content="https://bubbleboy11.github.io/2021/07/06/go-cuo-wu-he-yi-chang-panic-he-recover/index.html">
<meta property="og:site_name" content="外心人D的博客">
<meta property="og:description" content="错误和异常是两个不同的概念，非常容易混淆。很多程序员习惯将一切非正常情况都看做错误，一切皆错误或一切皆异常的情况，而不区分错误和异常，即使程序中可能有异常抛出，也将异常及时捕获并转换成错误。从表面上看，一切皆错误的思路更简单，而异常的引入仅仅增加了额外的复杂度。但事实并非如此。众所周知，Golang遵循“少即是多”的设计哲学，追求简洁优雅，就是说如果异常价值不大，就不会将异常加入到语言特性中。 错">
<meta property="og:locale" content="zh_CN">
<meta property="og:image" content="https://bubbleboy11.github.io/2021/07/06/go-cuo-wu-he-yi-chang-panic-he-recover/go%E7%9A%84http/1.png">
<meta property="article:published_time" content="2021-07-05T16:21:41.000Z">
<meta property="article:modified_time" content="2021-08-17T02:53:36.427Z">
<meta property="article:author" content="外心人D">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="https://bubbleboy11.github.io/2021/07/06/go-cuo-wu-he-yi-chang-panic-he-recover/go%E7%9A%84http/1.png">


<link rel="canonical" href="https://bubbleboy11.github.io/2021/07/06/go-cuo-wu-he-yi-chang-panic-he-recover/">



<script class="next-config" data-name="page" type="application/json">{"sidebar":"","isHome":false,"isPost":true,"lang":"zh-CN","comments":true,"permalink":"https://bubbleboy11.github.io/2021/07/06/go-cuo-wu-he-yi-chang-panic-he-recover/","path":"2021/07/06/go-cuo-wu-he-yi-chang-panic-he-recover/","title":"go错误和异常panic和recover"}</script>

<script class="next-config" data-name="calendar" type="application/json">""</script>
<title>go错误和异常panic和recover | 外心人D的博客</title>
  




  <noscript>
    <link rel="stylesheet" href="/css/noscript.css">
  </noscript>
<link rel="alternate" href="/atom.xml" title="外心人D的博客" type="application/atom+xml">
<link rel="stylesheet" href="/css/prism.css" type="text/css"></head>

<body itemscope itemtype="http://schema.org/WebPage" class="use-motion">
  <div class="headband"></div>

  <main class="main">
    <header class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-container">
  <div class="site-nav-toggle">
    <div class="toggle" aria-label="切换导航栏" role="button">
    </div>
  </div>

  <div class="site-meta">

    <a href="/" class="brand" rel="start">
      <i class="logo-line"></i>
      <h1 class="site-title">外心人D的博客</h1>
      <i class="logo-line"></i>
    </a>
  </div>

  <div class="site-nav-right">
    <div class="toggle popup-trigger">
    </div>
  </div>
</div>







</div>
        
  
  <div class="toggle sidebar-toggle" role="button">
    <span class="toggle-line"></span>
    <span class="toggle-line"></span>
    <span class="toggle-line"></span>
  </div>

  <aside class="sidebar">

    <div class="sidebar-inner sidebar-nav-active sidebar-toc-active">
      <ul class="sidebar-nav">
        <li class="sidebar-nav-toc">
          文章目录
        </li>
        <li class="sidebar-nav-overview">
          站点概览
        </li>
      </ul>

      <div class="sidebar-panel-container">
        <!--noindex-->
        <div class="post-toc-wrap sidebar-panel">
            <div class="post-toc animated"><ol class="nav"><li class="nav-item nav-level-2"><a class="nav-link" href="#%E5%AF%B9%E5%BE%85%E5%BC%82%E5%B8%B8%E7%9A%84%E6%80%81%E5%BA%A6"><span class="nav-text">对待异常的态度</span></a></li><li class="nav-item nav-level-2"><a class="nav-link" href="#1-%E5%9C%A8%E7%A8%8B%E5%BA%8F%E5%BC%80%E5%8F%91%E9%98%B6%E6%AE%B5%EF%BC%8C%E5%9D%9A%E6%8C%81%E9%80%9F%E9%94%99"><span class="nav-text">1. 在程序开发阶段，坚持速错</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#2-%E5%9C%A8%E7%A8%8B%E5%BA%8F%E9%83%A8%E7%BD%B2%E5%90%8E%EF%BC%8C%E5%BA%94%E6%81%A2%E5%A4%8D%E5%BC%82%E5%B8%B8%E9%81%BF%E5%85%8D%E7%A8%8B%E5%BA%8F%E7%BB%88%E6%AD%A2"><span class="nav-text">2. 在程序部署后，应恢复异常避免程序终止</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#3-%E5%AF%B9%E4%BA%8E%E4%B8%8D%E5%BA%94%E8%AF%A5%E5%87%BA%E7%8E%B0%E7%9A%84%E5%88%86%E6%94%AF%EF%BC%8C%E4%BD%BF%E7%94%A8%E5%BC%82%E5%B8%B8%E5%A4%84%E7%90%86"><span class="nav-text">3. 对于不应该出现的分支，使用异常处理</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#4-%E9%92%88%E5%AF%B9%E5%85%A5%E5%8F%82%E4%B8%8D%E5%BA%94%E8%AF%A5%E6%9C%89%E9%97%AE%E9%A2%98%E7%9A%84%E5%87%BD%E6%95%B0%EF%BC%8C%E4%BD%BF%E7%94%A8panic%E8%AE%BE%E8%AE%A1"><span class="nav-text">4. 针对入参不应该有问题的函数，使用panic设计</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#%E9%94%99%E8%AF%AF%E5%A4%84%E7%90%86%E7%9A%84%E6%AD%A3%E7%A1%AE%E5%A7%BF%E5%8A%BF"><span class="nav-text">错误处理的正确姿势</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#1-%E5%A4%B1%E8%B4%A5%E7%9A%84%E5%8E%9F%E5%9B%A0%E5%8F%AA%E6%9C%89%E4%B8%80%E4%B8%AA%E6%97%B6%EF%BC%8C%E4%B8%8D%E4%BD%BF%E7%94%A8error"><span class="nav-text">1. 失败的原因只有一个时，不使用error</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#2-%E6%B2%A1%E6%9C%89%E5%A4%B1%E8%B4%A5%E6%97%B6%EF%BC%8C%E4%B8%8D%E4%BD%BF%E7%94%A8error"><span class="nav-text">2. 没有失败时，不使用error</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#3-error%E5%BA%94%E6%94%BE%E5%9C%A8%E8%BF%94%E5%9B%9E%E5%80%BC%E7%B1%BB%E5%9E%8B%E5%88%97%E8%A1%A8%E7%9A%84%E6%9C%80%E5%90%8E"><span class="nav-text">3. error应放在返回值类型列表的最后</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#4-%E9%94%99%E8%AF%AF%E5%80%BC%E7%BB%9F%E4%B8%80%E5%AE%9A%E4%B9%89%EF%BC%8C%E8%80%8C%E4%B8%8D%E6%98%AF%E8%B7%9F%E7%9D%80%E6%84%9F%E8%A7%89%E8%B5%B0"><span class="nav-text">4. 错误值统一定义，而不是跟着感觉走</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#7-%E5%BD%93%E5%8F%91%E7%94%9F%E9%94%99%E8%AF%AF%E6%97%B6%EF%BC%8C%E4%B8%8D%E5%BF%BD%E7%95%A5%E6%9C%89%E7%94%A8%E7%9A%84%E8%BF%94%E5%9B%9E%E5%80%BC"><span class="nav-text">7. 当发生错误时，不忽略有用的返回值</span></a></li></ol></li><li class="nav-item nav-level-2"><a class="nav-link" href="#Golang%E9%94%99%E8%AF%AF%E5%92%8C%E5%BC%82%E5%B8%B8%E6%98%AF%E5%8F%AF%E4%BB%A5%E4%BA%92%E7%9B%B8%E8%BD%AC%E6%8D%A2%E7%9A%84%EF%BC%9A"><span class="nav-text">Golang错误和异常是可以互相转换的：</span></a></li></ol></div>
        </div>
        <!--/noindex-->

        <div class="site-overview-wrap sidebar-panel">
          <div class="site-overview">
            <div class="site-author site-overview-item animated" itemprop="author" itemscope itemtype="http://schema.org/Person">
  <p class="site-author-name" itemprop="name">外心人D</p>
  <div class="site-description" itemprop="description">本博客大多内容为慕课和网上博客，并非原创</div>
</div>
<div class="site-state-wrap site-overview-item animated">
  <nav class="site-state">
      <div class="site-state-item site-state-posts">
          <a href="/archives">
          <span class="site-state-item-count">343</span>
          <span class="site-state-item-name">日志</span>
        </a>
      </div>
      <div class="site-state-item site-state-categories">
            <a href="/categories/">
        <span class="site-state-item-count">22</span>
        <span class="site-state-item-name">分类</span></a>
      </div>
      <div class="site-state-item site-state-tags">
            <a href="/tags/">
        <span class="site-state-item-count">39</span>
        <span class="site-state-item-name">标签</span></a>
      </div>
  </nav>
</div>



          </div>
        </div>
      </div>
        <div class="back-to-top animated" role="button" aria-label="返回顶部">
          <i class="fa fa-arrow-up"></i>
          <span>0%</span>
        </div>
    </div>
  </aside>
  <div class="sidebar-dimmer"></div>


    </header>

    

<noscript>
  <div class="noscript-warning">Theme NexT works best with JavaScript enabled</div>
</noscript>


    <div class="main-inner post posts-expand">


  


<div class="post-block">
  
  

  <article itemscope itemtype="http://schema.org/Article" class="post-content" lang="zh-CN">
    <link itemprop="mainEntityOfPage" href="https://bubbleboy11.github.io/2021/07/06/go-cuo-wu-he-yi-chang-panic-he-recover/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/images/avatar.gif">
      <meta itemprop="name" content="外心人D">
      <meta itemprop="description" content="本博客大多内容为慕课和网上博客，并非原创">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="外心人D的博客">
    </span>
      <header class="post-header">
        <h1 class="post-title" itemprop="name headline">
          go错误和异常panic和recover
        </h1>

        <div class="post-meta-container">
          <div class="post-meta">
    <span class="post-meta-item">
      <span class="post-meta-item-icon">
        <i class="far fa-calendar"></i>
      </span>
      <span class="post-meta-item-text">发表于</span>

      <time title="创建时间：2021-07-05 16:21:41" itemprop="dateCreated datePublished" datetime="2021-07-05T16:21:41Z">2021-07-05</time>
    </span>
      <span class="post-meta-item">
        <span class="post-meta-item-icon">
          <i class="far fa-calendar-check"></i>
        </span>
        <span class="post-meta-item-text">更新于</span>
        <time title="修改时间：2021-08-17 02:53:36" itemprop="dateModified" datetime="2021-08-17T02:53:36Z">2021-08-17</time>
      </span>
    <span class="post-meta-item">
      <span class="post-meta-item-icon">
        <i class="far fa-folder"></i>
      </span>
      <span class="post-meta-item-text">分类于</span>
        <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
          <a href="/categories/go/" itemprop="url" rel="index"><span itemprop="name">go</span></a>
        </span>
    </span>

  
</div>

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">
        <p>错误和异常是两个不同的概念，非常容易混淆。很多程序员习惯将一切非正常情况都看做错误，<br>一切皆错误或一切皆异常的情况，而不区分错误和异常，即使程序中可能有异常抛出，也将异常及时捕获并转换成错误。从表面上看，一切皆错误的思路更简单，而异常的引入仅仅增加了额外的复杂度。<br>但事实并非如此。众所周知，Golang遵循“少即是多”的设计哲学，追求简洁优雅，就是说如果异常价值不大，就不会将异常加入到语言特性中。</p>
<p>错误和异常处理是程序的重要组成部分，我们先看看下面几个问题：</p>
<p>错误和异常如何区分？<br>错误和异常从Golang机制上讲，就是error和panic的区别。</p>
<p>错误处理的方式有哪几种？</p>
<p>什么时候需要使用异常终止程序？异常处理的作用域（场景）：</p>
<ul>
<li>空指针引用</li>
<li>下标越界</li>
<li>除数为0</li>
<li>不应该出现的分支，比如default</li>
<li>输入不应该引起函数错误</li>
</ul>
<p>其他场景我们使用错误处理，这使得我们的函数接口很精炼。</p>
<p>什么时候需要捕获异常？</p>
<p>Golang中引入两个内置函数 panic 和 recover 来触发**和终止异常处理流程，同时引入关键字defer来延迟执行defer后面的函数。<br>一直等到包含defer语句的函数执行完毕时，延迟函数（defer后的函数）才会被执行，而不管包含defer语句的函数是通过return的正常结束，还是由于panic导致的异常结束。</p>
<figure class="highlight go"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">// The panic built-in function stops normal execution of the current</span></span><br><span class="line"><span class="comment">// goroutine. When a function F calls panic, normal execution of F stops</span></span><br><span class="line"><span class="comment">// immediately. Any functions whose execution was deferred by F are run in</span></span><br><span class="line"><span class="comment">// the usual way, and then F returns to its caller. To the caller G, the</span></span><br><span class="line"><span class="comment">// invocation of F then behaves like a call to panic, terminating G's</span></span><br><span class="line"><span class="comment">// execution and running any deferred functions. This continues until all</span></span><br><span class="line"><span class="comment">// functions in the executing goroutine have stopped, in reverse order. At</span></span><br><span class="line"><span class="comment">// that point, the program is terminated with a non-zero exit code. This</span></span><br><span class="line"><span class="comment">// termination sequence is called panicking and can be controlled by the</span></span><br><span class="line"><span class="comment">// built-in function recover.</span></span><br><span class="line"><span class="function"><span class="keyword">func</span> <span class="title">panic</span><span class="params">(v <span class="keyword">interface</span>{})</span></span></span><br><span class="line"></span><br><span class="line"><span class="comment">// The recover built-in function allows a program to manage behavior of a</span></span><br><span class="line"><span class="comment">// panicking goroutine. Executing a call to recover inside a deferred</span></span><br><span class="line"><span class="comment">// function (but not any function called by it) stops the panicking sequence</span></span><br><span class="line"><span class="comment">// by restoring normal execution and retrieves the error value passed to the</span></span><br><span class="line"><span class="comment">// call of panic. If recover is called outside the deferred function it will</span></span><br><span class="line"><span class="comment">// not stop a panicking sequence. In this case, or when the goroutine is not</span></span><br><span class="line"><span class="comment">// panicking, or if the argument supplied to panic was nil, recover returns</span></span><br><span class="line"><span class="comment">// nil. Thus the return value from recover reports whether the goroutine is</span></span><br><span class="line"><span class="comment">// panicking.</span></span><br><span class="line"><span class="function"><span class="keyword">func</span> <span class="title">recover</span><span class="params">()</span> <span class="title">interface</span></span>{}</span><br></pre></td></tr></tbody></table></figure>

<p>当有的错误属于异常情况下，正常代码运行不会出现的问题的情况就需要用panic了，但是也需要进行panic的recover处理</p>
<p>意料之外使用panic</p>
<ul>
<li>不要经常用panic，和其他语言的throw差不多 比如C++/Java，没有error但有errno，没有panic但有throw。</li>
<li>停止当前函数的执行</li>
<li>一直向上返回，执行每一层的defer</li>
<li>如果没有遇见recover，程序退出</li>
</ul>
<p>recover</p>
<ul>
<li>仅在 defer 调用中使用</li>
<li>获取 panic 的值</li>
<li>如果无法处理，可以重新 panic</li>
</ul>
<p>当程序运行时，如果遇到引用空指针、下标越界或显式调用panic函数等情况，则先触发panic函数的执行，然后调用延迟函数。调用者继续传递panic，因此该过程一直在调用栈中重复发生：函数停止执行，调用延迟执行函数等。如果一路在延迟函数中没有recover函数的调用，则会到达该携程的起点，该携程结束，然后终止其他所有携程，包括主携程（类似于C语言中的主线程，该携程ID为1）。</p>
<p>对于异常，我们可以选择在一个合适的上游去recover，并打印堆栈信息，使得部署后的程序不会终止。</p>
<h2 id="对待异常的态度"><a href="#对待异常的态度" class="headerlink" title="对待异常的态度"></a>对待异常的态度</h2><h2 id="1-在程序开发阶段，坚持速错"><a href="#1-在程序开发阶段，坚持速错" class="headerlink" title="1. 在程序开发阶段，坚持速错"></a>1. 在程序开发阶段，坚持速错</h2><p>去年学习Erlang的时候，建立了速错的理念，简单来讲就是“让它挂”，只有挂了你才会第一时间知道错误。在早期开发以及任何发布阶段之前，最简单的同时也可能是最好的方法是调用panic函数来中断程序的执行以强制发生错误，使得该错误不会被忽略，因而能够被尽快修复。</p>
<h3 id="2-在程序部署后，应恢复异常避免程序终止"><a href="#2-在程序部署后，应恢复异常避免程序终止" class="headerlink" title="2. 在程序部署后，应恢复异常避免程序终止"></a>2. 在程序部署后，应恢复异常避免程序终止</h3><p>我们在调用recover的延迟函数中以最合理的方式响应该异常：</p>
<ol>
<li>打印堆栈的异常调用信息和关键的业务信息，以便这些问题保留可见；</li>
<li>将异常转换为错误，以便调用者让程序恢复到健康状态并继续安全运行。</li>
</ol>
<figure class="highlight go"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="keyword">func</span> <span class="title">funcA</span><span class="params">()</span> <span class="title">error</span></span> {</span><br><span class="line">    <span class="keyword">defer</span> <span class="function"><span class="keyword">func</span><span class="params">()</span></span> {</span><br><span class="line">        <span class="keyword">if</span> p := <span class="built_in">recover</span>(); p != <span class="literal">nil</span> {</span><br><span class="line">            fmt.Printf(<span class="string">"panic recover! p: %v"</span>, p)</span><br><span class="line">            debug.PrintStack()</span><br><span class="line">        }</span><br><span class="line">    }()</span><br><span class="line">    <span class="keyword">return</span> funcB()</span><br><span class="line">}</span><br><span class="line"><span class="function"><span class="keyword">func</span> <span class="title">funcB</span><span class="params">()</span> <span class="title">error</span></span> {</span><br><span class="line">    <span class="comment">// simulation</span></span><br><span class="line">    <span class="built_in">panic</span>(<span class="string">"foo"</span>)</span><br><span class="line">    <span class="keyword">return</span> errors.New(<span class="string">"success"</span>)</span><br><span class="line">}</span><br><span class="line"><span class="function"><span class="keyword">func</span> <span class="title">test</span><span class="params">()</span></span> {</span><br><span class="line">    err := funcA()</span><br><span class="line">    <span class="keyword">if</span> err == <span class="literal">nil</span> {</span><br><span class="line">        fmt.Printf(<span class="string">"err is nil\\n"</span>)</span><br><span class="line">    } <span class="keyword">else</span> {</span><br><span class="line">        fmt.Printf(<span class="string">"err is %v\\n"</span>, err)</span><br><span class="line">    }</span><br><span class="line">}</span><br></pre></td></tr></tbody></table></figure>

<p>我们期望test函数的输出是：</p>
<figure class="highlight text"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">err is foo</span><br></pre></td></tr></tbody></table></figure>

<p>实际上test函数的输出是：</p>
<figure class="highlight text"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">err is nil</span><br></pre></td></tr></tbody></table></figure>

<p>原因是panic异常处理机制不会自动将错误信息传递给error，所以要在funcA函数中进行显式的传递，代码如下所示：</p>
<figure class="highlight go"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="keyword">func</span> <span class="title">funcA</span><span class="params">()</span> <span class="params">(err error)</span></span> {</span><br><span class="line">    <span class="keyword">defer</span> <span class="function"><span class="keyword">func</span><span class="params">()</span></span> {</span><br><span class="line">        <span class="keyword">if</span> p := <span class="built_in">recover</span>(); p != <span class="literal">nil</span> {</span><br><span class="line">            fmt.Println(<span class="string">"panic recover! p:"</span>, p)</span><br><span class="line">            str, ok := p.(<span class="keyword">string</span>)</span><br><span class="line">            <span class="keyword">if</span> ok {</span><br><span class="line">                err = errors.New(str)</span><br><span class="line">            } <span class="keyword">else</span> {</span><br><span class="line">                err = errors.New(<span class="string">"panic"</span>)</span><br><span class="line">            }</span><br><span class="line">            debug.PrintStack()</span><br><span class="line">        }</span><br><span class="line">    }()</span><br><span class="line">    <span class="keyword">return</span> funcB()</span><br><span class="line">}</span><br></pre></td></tr></tbody></table></figure>

<figure class="highlight go"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="keyword">func</span> <span class="title">tryRecover</span><span class="params">()</span></span> {</span><br><span class="line">    <span class="keyword">defer</span> <span class="function"><span class="keyword">func</span><span class="params">()</span></span> { <span class="comment">// func(){} 只是一个函数体， 要运行这个函数要在后面加上()</span></span><br><span class="line">        r := <span class="built_in">recover</span>() <span class="comment">// recover只能在defer中调用</span></span><br><span class="line">        <span class="keyword">if</span> r == <span class="literal">nil</span> { <span class="comment">// 如果没有异常  r为nil</span></span><br><span class="line">            fmt.Println(<span class="string">"Nothing to recover. "</span> + <span class="string">"Please try uncomment errors "</span> + <span class="string">"below."</span>)</span><br><span class="line">            <span class="keyword">return</span></span><br><span class="line">        }</span><br><span class="line">        <span class="keyword">if</span> err, ok := r.(error); ok {</span><br><span class="line">            fmt.Println(<span class="string">"Error occurred:"</span>, err)</span><br><span class="line">            fmt.Println(<span class="string">"Error occurred:"</span>, err, err.Error())</span><br><span class="line">        } <span class="keyword">else</span> {</span><br><span class="line">            <span class="comment">// 不知道怎么处理 可以继续panic</span></span><br><span class="line">            <span class="built_in">panic</span>(fmt.Sprintf(<span class="string">"I don't know what to do: %v"</span>, r))</span><br><span class="line">        }</span><br><span class="line">    }()</span><br><span class="line"></span><br><span class="line">    <span class="comment">// Uncomment each block to see different panic scenarios.</span></span><br><span class="line">    <span class="comment">// Normal error</span></span><br><span class="line">    <span class="comment">//panic(errors.New("this is an error"))</span></span><br><span class="line">    <span class="comment">// Error occurred: this is an error</span></span><br><span class="line"></span><br><span class="line">    <span class="comment">// Division by zero</span></span><br><span class="line">    <span class="comment">//b := 0</span></span><br><span class="line">    <span class="comment">//a := 5 / b</span></span><br><span class="line">    <span class="comment">//fmt.Println(a)</span></span><br><span class="line">    <span class="comment">// Error occurred: runtime error: integer divide by zero</span></span><br><span class="line"></span><br><span class="line">    <span class="comment">// Causes re-panic</span></span><br><span class="line">    <span class="built_in">panic</span>(<span class="number">123</span>) <span class="comment">// 这个因为不是错误 defer中会继续抛出</span></span><br><span class="line">    <span class="comment">// panic: 123 [recovered]</span></span><br><span class="line">    <span class="comment">//     panic: I don't know what to do: 123</span></span><br><span class="line">}</span><br><span class="line"></span><br><span class="line"><span class="function"><span class="keyword">func</span> <span class="title">main</span><span class="params">()</span></span> {</span><br><span class="line">    tryRecover()</span><br><span class="line">}</span><br></pre></td></tr></tbody></table></figure>

<h3 id="3-对于不应该出现的分支，使用异常处理"><a href="#3-对于不应该出现的分支，使用异常处理" class="headerlink" title="3. 对于不应该出现的分支，使用异常处理"></a>3. 对于不应该出现的分支，使用异常处理</h3><p>当某些不应该发生的场景发生时，我们就应该调用panic函数来触发异常。比如，当程序到达了某条逻辑上不可能到达的路径：</p>
<figure class="highlight go"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">switch</span> s := suit(drawCard()); s {</span><br><span class="line">    <span class="keyword">case</span> <span class="string">"Spades"</span>:</span><br><span class="line">    <span class="comment">// ...</span></span><br><span class="line">    <span class="keyword">case</span> <span class="string">"Hearts"</span>:</span><br><span class="line">    <span class="comment">// ...</span></span><br><span class="line">    <span class="keyword">case</span> <span class="string">"Diamonds"</span>:</span><br><span class="line">    <span class="comment">// ...</span></span><br><span class="line">    <span class="keyword">case</span> <span class="string">"Clubs"</span>:</span><br><span class="line">    <span class="comment">// ...</span></span><br><span class="line">    <span class="keyword">default</span>:</span><br><span class="line">        <span class="built_in">panic</span>(fmt.Sprintf(<span class="string">"invalid suit %v"</span>, s))</span><br><span class="line">}</span><br></pre></td></tr></tbody></table></figure>

<h3 id="4-针对入参不应该有问题的函数，使用panic设计"><a href="#4-针对入参不应该有问题的函数，使用panic设计" class="headerlink" title="4. 针对入参不应该有问题的函数，使用panic设计"></a>4. 针对入参不应该有问题的函数，使用panic设计</h3><p>入参不应该有问题一般指的是硬编码，我们先看“一个启示”一节中提到的两个函数（Compile和MustCompile），其中MustCompile函数是对Compile函数的包装：</p>
<figure class="highlight go"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="keyword">func</span> <span class="title">MustCompile</span><span class="params">(str <span class="keyword">string</span>)</span> *<span class="title">Regexp</span></span> {</span><br><span class="line">    regexp, error := Compile(str)</span><br><span class="line">    <span class="keyword">if</span> error != <span class="literal">nil</span> {</span><br><span class="line">        <span class="built_in">panic</span>(<span class="string">`regexp: Compile(`</span> + quote(str) + <span class="string">`): `</span> + error.Error())</span><br><span class="line">    }</span><br><span class="line">    <span class="keyword">return</span> regexp</span><br><span class="line">}</span><br></pre></td></tr></tbody></table></figure>

<p>所以，对于同时支持用户输入场景和硬编码场景的情况，一般支持硬编码场景的函数是对支持用户输入场景函数的包装。<br>对于只支持硬编码单一场景的情况，函数设计时直接使用panic，即返回值类型列表中不会有error，这使得函数的调用处理非常方便（没有了乏味的”if err != nil {/ 打印 &amp;&amp; 错误处理 /}”代码块）。</p>
<figure class="highlight go"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="keyword">func</span> <span class="title">div</span><span class="params">(a, b <span class="keyword">int</span>)</span> <span class="params">(<span class="keyword">int</span>, error)</span></span> {</span><br><span class="line"></span><br><span class="line">    <span class="keyword">return</span> a / b, <span class="literal">nil</span></span><br><span class="line">}</span><br><span class="line"></span><br><span class="line"><span class="function"><span class="keyword">func</span> <span class="title">main</span><span class="params">()</span></span> {</span><br><span class="line">    <span class="comment">//错误就是能遇到可能出现的情况，这些情况会导致你的代码出问题， 参数检查， 数据库访问不了</span></span><br><span class="line">    a := <span class="number">12</span></span><br><span class="line">    b := <span class="number">0</span></span><br><span class="line">    fmt.Println(div(a, b)) <span class="comment">// panic: runtime error: integer divide by zero</span></span><br><span class="line">}</span><br></pre></td></tr></tbody></table></figure>

<figure class="highlight go"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="keyword">func</span> <span class="title">div1</span><span class="params">(a, b <span class="keyword">int</span>)</span> <span class="params">(<span class="keyword">int</span>, error)</span></span> {</span><br><span class="line">    <span class="keyword">if</span> b == <span class="number">0</span> {</span><br><span class="line">        <span class="built_in">panic</span>(<span class="string">"被除数不能为0"</span>)</span><br><span class="line">    }</span><br><span class="line">    <span class="keyword">return</span> a / b, <span class="literal">nil</span></span><br><span class="line">}</span><br><span class="line"></span><br><span class="line"><span class="function"><span class="keyword">func</span> <span class="title">main</span><span class="params">()</span></span> {</span><br><span class="line">    <span class="comment">//错误就是能遇到可能出现的情况，这些情况会导致你的代码出问题， 参数检查， 数据库访问不了</span></span><br><span class="line">    a := <span class="number">12</span></span><br><span class="line">    b := <span class="number">0</span></span><br><span class="line">    fmt.Println(div1(a, b)) <span class="comment">// panic: 被除数不能为0</span></span><br><span class="line">}</span><br></pre></td></tr></tbody></table></figure>

<figure class="highlight go"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="keyword">func</span> <span class="title">div1</span><span class="params">(a, b <span class="keyword">int</span>)</span> <span class="params">(<span class="keyword">int</span>, error)</span></span> {</span><br><span class="line">    <span class="keyword">if</span> b == <span class="number">0</span> {</span><br><span class="line">        <span class="built_in">panic</span>(<span class="string">"被除数不能为0"</span>)</span><br><span class="line">    }</span><br><span class="line">    <span class="keyword">return</span> a / b, <span class="literal">nil</span></span><br><span class="line">}</span><br><span class="line"></span><br><span class="line"><span class="function"><span class="keyword">func</span> <span class="title">main</span><span class="params">()</span></span> {</span><br><span class="line">    <span class="comment">//错误就是能遇到可能出现的情况，这些情况会导致你的代码出问题， 参数检查， 数据库访问不了</span></span><br><span class="line">    a := <span class="number">12</span></span><br><span class="line">    b := <span class="number">0</span></span><br><span class="line">    <span class="comment">// 出现错误函数不会终止，捕获异常</span></span><br><span class="line">    <span class="keyword">defer</span> <span class="function"><span class="keyword">func</span><span class="params">()</span></span> {</span><br><span class="line">       err := <span class="built_in">recover</span>()</span><br><span class="line">       <span class="keyword">if</span> err != <span class="literal">nil</span> {</span><br><span class="line">           fmt.Println(<span class="string">"异常被捕获到"</span>)</span><br><span class="line">       }</span><br><span class="line">       fmt.Println(<span class="string">"bobby"</span>)</span><br><span class="line">    }()</span><br><span class="line">    fmt.Println(div1(a, b))</span><br><span class="line">    <span class="comment">// 异常被捕获到</span></span><br><span class="line">    <span class="comment">// bobby</span></span><br><span class="line">}</span><br></pre></td></tr></tbody></table></figure>

<figure class="highlight go"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"><span class="string">``</span><span class="string">`</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">panic的坑</span></span><br><span class="line"><span class="string">异常 go语言中如何抛出异常和如何捕捉异常</span></span><br><span class="line"><span class="string">go语言认为错误就要自己处理， 就个人而言，go语言的这种想法是正确的。但是实际的使用中确实人有点烦人</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">`</span><span class="string">``</span><span class="keyword">go</span></span><br><span class="line">    strconv.Itoa(data)             <span class="comment">//go 认为这个Itoa的函数不可能出错， 没有必要返回error 内部代码出错这个时候应该抛出异常 panic</span></span><br><span class="line">    _, err := strconv.Atoi(<span class="string">"abcd"</span>) <span class="comment">//Atoi这个函数认为我的函数内部会出现一些预知错误情况</span></span><br><span class="line">    <span class="keyword">if</span> err != <span class="literal">nil</span> {</span><br><span class="line">        <span class="comment">//错误</span></span><br><span class="line">    }</span><br></pre></td></tr></tbody></table></figure>

<figure class="highlight go"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="keyword">func</span> <span class="title">main</span><span class="params">()</span></span> {</span><br><span class="line">    http.HandleFunc(<span class="string">"/hello"</span>, <span class="function"><span class="keyword">func</span><span class="params">(w http.ResponseWriter, r *http.Request)</span></span> {</span><br><span class="line">        w.Write([]<span class="keyword">byte</span>(<span class="string">"hello world!"</span>))</span><br><span class="line">    })</span><br><span class="line"></span><br><span class="line">    http.ListenAndServe(<span class="string">"127.0.0.1:8080"</span>, <span class="literal">nil</span>)</span><br><span class="line">}</span><br></pre></td></tr></tbody></table></figure>

<p><img src="./go%E7%9A%84http/1.png" alt="description"></p>
<figure class="highlight go"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="keyword">func</span> <span class="title">main</span><span class="params">()</span></span> {</span><br><span class="line">    http.HandleFunc(<span class="string">"/hello"</span>, <span class="function"><span class="keyword">func</span><span class="params">(w http.ResponseWriter, r *http.Request)</span></span> {</span><br><span class="line">        a := <span class="number">12</span></span><br><span class="line">        b := <span class="number">0</span></span><br><span class="line">        fmt.Println(div(a, b)) <span class="comment">// 浏览器刷新 http: panic serving 127.0.0.1:53676: runtime error: integer divide by zero</span></span><br><span class="line">        <span class="comment">// 出现异常这个程序没有挂掉，一个协程挂掉不会影响整体服务的运行</span></span><br><span class="line">        w.Write([]<span class="keyword">byte</span>(<span class="string">"hello world!"</span>))</span><br><span class="line">    })</span><br><span class="line"></span><br><span class="line">    http.ListenAndServe(<span class="string">"127.0.0.1:8080"</span>, <span class="literal">nil</span>)</span><br><span class="line">}</span><br></pre></td></tr></tbody></table></figure>

<figure class="highlight go"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="keyword">func</span> <span class="title">main</span><span class="params">()</span></span> {</span><br><span class="line">    http.HandleFunc(<span class="string">"/hello"</span>, <span class="function"><span class="keyword">func</span><span class="params">(w http.ResponseWriter, r *http.Request)</span></span> {</span><br><span class="line">        <span class="built_in">panic</span>(<span class="string">"error"</span>)</span><br><span class="line">        <span class="comment">// http: panic serving 127.0.0.1:59184: error</span></span><br><span class="line">        w.Write([]<span class="keyword">byte</span>(<span class="string">"hello world!"</span>))</span><br><span class="line">    })</span><br><span class="line">    http.ListenAndServe(<span class="string">"127.0.0.1:8080"</span>, <span class="literal">nil</span>)</span><br><span class="line">}</span><br></pre></td></tr></tbody></table></figure>

<p>panic会引起主线程的挂掉， 同时会导致其他的协程都挂掉<br>在父协程中无法捕获子协程中出现的异常</p>
<figure class="highlight go"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="keyword">func</span> <span class="title">main</span><span class="params">()</span></span> {</span><br><span class="line">    http.HandleFunc(<span class="string">"/hello"</span>, <span class="function"><span class="keyword">func</span><span class="params">(w http.ResponseWriter, r *http.Request)</span></span> {</span><br><span class="line">        <span class="keyword">go</span> <span class="function"><span class="keyword">func</span><span class="params">()</span></span> { <span class="comment">// 开启一个协程</span></span><br><span class="line">           <span class="built_in">panic</span>(<span class="string">"error"</span>) <span class="comment">// panic: error</span></span><br><span class="line">        }()</span><br><span class="line">        <span class="comment">//操作redis的，有人觉得这段代码可以放到协程中去运行。有一个非常大的隐患了</span></span><br><span class="line">        w.Write([]<span class="keyword">byte</span>(<span class="string">"hello world!"</span>))</span><br><span class="line">    })</span><br><span class="line"></span><br><span class="line">    http.ListenAndServe(<span class="string">"127.0.0.1:8080"</span>, <span class="literal">nil</span>)</span><br><span class="line">}</span><br></pre></td></tr></tbody></table></figure>

<figure class="highlight go"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="keyword">func</span> <span class="title">f1</span><span class="params">()</span></span> {</span><br><span class="line">    <span class="keyword">defer</span> <span class="function"><span class="keyword">func</span><span class="params">()</span></span> {</span><br><span class="line">        err := <span class="built_in">recover</span>()</span><br><span class="line">        <span class="keyword">if</span> err != <span class="literal">nil</span> {</span><br><span class="line">            fmt.Println(<span class="string">"捕获到了"</span>)</span><br><span class="line">        }</span><br><span class="line">    }()</span><br><span class="line"></span><br><span class="line">    <span class="built_in">panic</span>(<span class="string">"出错了"</span>) <span class="comment">// 主协程</span></span><br><span class="line">}</span><br><span class="line"></span><br><span class="line"><span class="function"><span class="keyword">func</span> <span class="title">main</span><span class="params">()</span></span> {</span><br><span class="line">    f1() <span class="comment">// 捕获到了</span></span><br><span class="line">}</span><br></pre></td></tr></tbody></table></figure>

<figure class="highlight go"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="keyword">func</span> <span class="title">f1</span><span class="params">()</span></span> {</span><br><span class="line">    <span class="keyword">defer</span> <span class="function"><span class="keyword">func</span><span class="params">()</span></span> {</span><br><span class="line">        err := <span class="built_in">recover</span>()</span><br><span class="line">        <span class="keyword">if</span> err != <span class="literal">nil</span> {</span><br><span class="line">            fmt.Println(<span class="string">"捕获到了"</span>)</span><br><span class="line">        }</span><br><span class="line">    }()</span><br><span class="line"></span><br><span class="line">    <span class="keyword">go</span> <span class="function"><span class="keyword">func</span><span class="params">()</span></span> { <span class="comment">// 开启一个协程</span></span><br><span class="line">        <span class="keyword">defer</span> <span class="function"><span class="keyword">func</span><span class="params">()</span></span> {</span><br><span class="line">            err := <span class="built_in">recover</span>() <span class="comment">// 处理异常，否则会导致其他的整个服务挂掉</span></span><br><span class="line">            <span class="keyword">if</span> err != <span class="literal">nil</span> {</span><br><span class="line">                fmt.Println(<span class="string">"捕获到了2"</span>)</span><br><span class="line">            }</span><br><span class="line">        }()</span><br><span class="line"></span><br><span class="line">        <span class="built_in">panic</span>(<span class="string">"出错了"</span>)</span><br><span class="line">    }()</span><br><span class="line">    time.Sleep(<span class="number">10</span> * time.Second) <span class="comment">// 主协程执行完不释放清除，防止主协程退出</span></span><br><span class="line">    <span class="comment">//panic("出错了")</span></span><br><span class="line">}</span><br><span class="line"></span><br><span class="line"><span class="function"><span class="keyword">func</span> <span class="title">main</span><span class="params">()</span></span> {</span><br><span class="line">    f1() <span class="comment">// 捕获到了2</span></span><br><span class="line">}</span><br></pre></td></tr></tbody></table></figure>

<h2 id="错误处理的正确姿势"><a href="#错误处理的正确姿势" class="headerlink" title="错误处理的正确姿势"></a>错误处理的正确姿势</h2><p>意料之中使用error<br>error是用于防止错误情况的问题，用error进行错误提示<br>error接口类型作为错误处理的标准模式，如果函数要返回错误，则返回值类型列表中肯定包含error。<br>error处理过程类似于C语言中的错误码，可逐层返回，直到被处理。</p>
<p>说明： Golang错误处理方式一直是很多人诟病的地方，有些人吐槽说一半的代码都是<code>"if err != nil { / 打印 &amp;&amp; 错误处理 / }"</code>，严重影响正常的处理逻辑。</p>
<figure class="highlight go"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">// The error built-in interface type is the conventional interface for</span></span><br><span class="line"><span class="comment">// representing an error condition, with the nil value representing no error.</span></span><br><span class="line"><span class="keyword">type</span> error <span class="keyword">interface</span> { <span class="comment">// error是一个接口，不能实例化</span></span><br><span class="line">    Error() <span class="keyword">string</span></span><br><span class="line">}</span><br></pre></td></tr></tbody></table></figure>

<figure class="highlight go"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">// New returns an error that formats as the given text.</span></span><br><span class="line"><span class="comment">// Each call to New returns a distinct error value even if the text is identical.</span></span><br><span class="line"><span class="function"><span class="keyword">func</span> <span class="title">New</span><span class="params">(text <span class="keyword">string</span>)</span> <span class="title">error</span></span> {</span><br><span class="line">    <span class="keyword">return</span> &amp;errorString{text}</span><br><span class="line">}</span><br></pre></td></tr></tbody></table></figure>

<figure class="highlight go"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">// Errorf formats according to a format specifier and returns the string as a</span></span><br><span class="line"><span class="comment">// value that satisfies error.</span></span><br><span class="line"><span class="comment">//</span></span><br><span class="line"><span class="comment">// If the format specifier includes a %w verb with an error operand,</span></span><br><span class="line"><span class="comment">// the returned error will implement an Unwrap method returning the operand. It is</span></span><br><span class="line"><span class="comment">// invalid to include more than one %w verb or to supply it with an operand</span></span><br><span class="line"><span class="comment">// that does not implement the error interface. The %w verb is otherwise</span></span><br><span class="line"><span class="comment">// a synonym for %v.</span></span><br><span class="line"><span class="function"><span class="keyword">func</span> <span class="title">Errorf</span><span class="params">(format <span class="keyword">string</span>, a ...<span class="keyword">interface</span>{})</span> <span class="title">error</span></span> {</span><br><span class="line">    p := newPrinter()</span><br><span class="line">    p.wrapErrs = <span class="literal">true</span></span><br><span class="line">    p.doPrintf(format, a)</span><br><span class="line">    s := <span class="keyword">string</span>(p.buf)</span><br><span class="line">    <span class="keyword">var</span> err error</span><br><span class="line">    <span class="keyword">if</span> p.wrappedErr == <span class="literal">nil</span> {</span><br><span class="line">        err = errors.New(s)</span><br><span class="line">    } <span class="keyword">else</span> {</span><br><span class="line">        err = &amp;wrapError{s, p.wrappedErr}</span><br><span class="line">    }</span><br><span class="line">    p.free()</span><br><span class="line">    <span class="keyword">return</span> err</span><br><span class="line">}</span><br></pre></td></tr></tbody></table></figure>

<figure class="highlight go"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">var</span> err error = errors.New(<span class="string">"错误"</span>)</span><br><span class="line">fmt.Println(err) <span class="comment">// 错误</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">var</span> err1 error = errors.New(fmt.Sprintf(<span class="string">""</span>)) <span class="comment">// 格式化字符串</span></span><br><span class="line">fmt.Println(err1)</span><br><span class="line"></span><br><span class="line">s := <span class="string">"文件不存在"</span></span><br><span class="line"><span class="keyword">var</span> err2 error = fmt.Errorf(<span class="string">"错误:%s"</span>, s) <span class="comment">// 格式化字符串</span></span><br><span class="line">fmt.Println(err2) <span class="comment">// 错误:文件不存在</span></span><br></pre></td></tr></tbody></table></figure>

<figure class="highlight go"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="keyword">func</span> <span class="title">writeFile</span><span class="params">(filename <span class="keyword">string</span>)</span></span> {</span><br><span class="line"></span><br><span class="line">    <span class="comment">// If there is an error, it will be of type *PathError.</span></span><br><span class="line">    file, err := os.OpenFile(filename,</span><br><span class="line">        os.O_EXCL|os.O_CREATE|os.O_WRONLY, <span class="number">0666</span>)</span><br><span class="line"></span><br><span class="line">    <span class="keyword">if</span> err != <span class="literal">nil</span> {</span><br><span class="line">        <span class="comment">// fmt.Println("error", err.Error()) //error open fib.txt: file exists</span></span><br><span class="line">        <span class="comment">// return</span></span><br><span class="line"></span><br><span class="line">        <span class="keyword">if</span> pathError, ok := err.(*os.PathError); !ok { <span class="comment">// 判断err是否为指定类型</span></span><br><span class="line">            fmt.Printf(<span class="string">"%s, %s, %s\n"</span>, pathError.Op, pathError.Path, pathError.Err)</span><br><span class="line">            <span class="built_in">panic</span>(err)</span><br><span class="line">        } <span class="keyword">else</span> {</span><br><span class="line">            fmt.Printf(<span class="string">"%s, %s, %s\n"</span>,</span><br><span class="line">                pathError.Op, pathError.Path, pathError.Err)</span><br><span class="line">        }</span><br><span class="line">        <span class="keyword">return</span></span><br><span class="line">    }</span><br><span class="line">    <span class="keyword">defer</span> file.Close()</span><br><span class="line"></span><br><span class="line">    writer := bufio.NewWriter(file)</span><br><span class="line">    <span class="keyword">defer</span> writer.Flush() <span class="comment">// 把缓存的数据刷到文件中去</span></span><br><span class="line"></span><br><span class="line">    f := fib.Fibonacci()</span><br><span class="line">    <span class="keyword">for</span> i := <span class="number">0</span>; i &lt; <span class="number">20</span>; i++ {</span><br><span class="line">        fmt.Fprintln(writer, f())</span><br><span class="line">    }</span><br><span class="line">}</span><br><span class="line"><span class="function"><span class="keyword">func</span> <span class="title">main</span><span class="params">()</span></span> {</span><br><span class="line">    writeFile(<span class="string">"fib.txt"</span>) <span class="comment">// open, fib.txt, The file exists.</span></span><br><span class="line">}</span><br></pre></td></tr></tbody></table></figure>

<h3 id="1-失败的原因只有一个时，不使用error"><a href="#1-失败的原因只有一个时，不使用error" class="headerlink" title="1. 失败的原因只有一个时，不使用error"></a>1. 失败的原因只有一个时，不使用error</h3><figure class="highlight go"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="keyword">func</span> <span class="params">(self *AgentContext)</span> <span class="title">CheckHostType</span><span class="params">(host_type <span class="keyword">string</span>)</span> <span class="title">error</span></span> {</span><br><span class="line">    <span class="keyword">switch</span> host_type {</span><br><span class="line">    <span class="keyword">case</span> <span class="string">"virtual_machine"</span>:</span><br><span class="line">        <span class="keyword">return</span> <span class="literal">nil</span></span><br><span class="line">    <span class="keyword">case</span> <span class="string">"bare_metal"</span>:</span><br><span class="line">        <span class="keyword">return</span> <span class="literal">nil</span></span><br><span class="line">    }</span><br><span class="line">    <span class="keyword">return</span> errors.New(<span class="string">"CheckHostType ERROR:"</span> + host_type)</span><br><span class="line">}</span><br></pre></td></tr></tbody></table></figure>

<p>我们可以看出，该函数失败的原因只有一个，所以返回值的类型应该为bool，而不是error，重构一下代码：</p>
<figure class="highlight go"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="keyword">func</span> <span class="params">(self *AgentContext)</span> <span class="title">IsValidHostType</span><span class="params">(hostType <span class="keyword">string</span>)</span> <span class="title">bool</span></span> {</span><br><span class="line">    <span class="keyword">return</span> hostType == <span class="string">"virtual_machine"</span> || hostType == <span class="string">"bare_metal"</span></span><br><span class="line">}</span><br></pre></td></tr></tbody></table></figure>

<p>说明：大多数情况，导致失败的原因不止一种，尤其是对I/O操作而言，用户需要了解更多的错误信息，这时的返回值类型不再是简单的bool，而是error。</p>
<h3 id="2-没有失败时，不使用error"><a href="#2-没有失败时，不使用error" class="headerlink" title="2. 没有失败时，不使用error"></a>2. 没有失败时，不使用error</h3><h3 id="3-error应放在返回值类型列表的最后"><a href="#3-error应放在返回值类型列表的最后" class="headerlink" title="3. error应放在返回值类型列表的最后"></a>3. error应放在返回值类型列表的最后</h3><h3 id="4-错误值统一定义，而不是跟着感觉走"><a href="#4-错误值统一定义，而不是跟着感觉走" class="headerlink" title="4. 错误值统一定义，而不是跟着感觉走"></a>4. 错误值统一定义，而不是跟着感觉走</h3><p>很多人写代码时，到处<code>return errors.New(value)</code>，而错误value在表达同一个含义时也可能形式不同，比如“记录不存在”的错误value可能为：</p>
<ol>
<li><code>"record is not existed."</code></li>
<li><code>"record is not exist!"</code></li>
<li><code>"###record is not existed！！！"</code></li>
<li>…<br>这使得相同的错误value撒在一大片代码里，当上层函数要对特定错误value进行统一处理时，需要漫游所有下层代码，以保证错误value统一，不幸的是有时会有漏网之鱼，而且这种方式严重阻碍了错误value的重构。<br>于是，我们可以参考C/C++的错误码定义文件，在Golang的每个包中增加一个错误对象定义文件，如下所示</li>
</ol>
<figure class="highlight go"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">var</span> ERR_EOF = errors.New(<span class="string">"EOF"</span>)</span><br><span class="line"><span class="keyword">var</span> ERR_CLOSED_PIPE = errors.New(<span class="string">"io: read/write on closed pipe"</span>)</span><br><span class="line"><span class="keyword">var</span> ERR_NO_PROGRESS = errors.New(<span class="string">"multiple Read calls return no data or error"</span>)</span><br><span class="line"><span class="keyword">var</span> ERR_SHORT_BUFFER = errors.New(<span class="string">"short buffer"</span>)</span><br><span class="line"><span class="keyword">var</span> ERR_SHORT_WRITE = errors.New(<span class="string">"short write"</span>)</span><br><span class="line"><span class="keyword">var</span> ERR_UNEXPECTED_EOF = errors.New(<span class="string">"unexpected EOF"</span>)</span><br></pre></td></tr></tbody></table></figure>

<figure class="highlight go"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">package</span> main</span><br><span class="line"></span><br><span class="line"><span class="keyword">import</span> (</span><br><span class="line">    <span class="string">"io/ioutil"</span></span><br><span class="line">    <span class="string">"net/http"</span></span><br><span class="line">    <span class="string">"os"</span></span><br><span class="line">)</span><br><span class="line"></span><br><span class="line"><span class="function"><span class="keyword">func</span> <span class="title">main</span><span class="params">()</span></span> {</span><br><span class="line">    http.HandleFunc(<span class="string">"/list/"</span>, <span class="function"><span class="keyword">func</span><span class="params">(writer http.ResponseWriter, request *http.Request)</span></span> {</span><br><span class="line">        path := request.URL.Path[<span class="built_in">len</span>(<span class="string">"/list/"</span>):]</span><br><span class="line">        file, err := os.Open(path)</span><br><span class="line">        <span class="keyword">if</span> err != <span class="literal">nil</span> {</span><br><span class="line">            <span class="comment">// 访问 ： http://localhost:8888/list/fib.txta</span></span><br><span class="line">            <span class="comment">// open fib.txta: no such file or directory</span></span><br><span class="line">            <span class="comment">// 这样会把我们服务器内部错误暴露给客户端，这样做不好</span></span><br><span class="line">            <span class="comment">// 而且这个本来是一个业务逻辑，不要把错误处理和业务逻辑耦合在一起，把业务逻辑和错误处理分开</span></span><br><span class="line">            http.Error(writer, err.Error(), http.StatusInternalServerError)</span><br><span class="line">            <span class="keyword">return</span></span><br><span class="line">        }</span><br><span class="line">        <span class="keyword">defer</span> file.Close()</span><br><span class="line"></span><br><span class="line">        all, err := ioutil.ReadAll(file)</span><br><span class="line">        <span class="keyword">if</span> err != <span class="literal">nil</span> {</span><br><span class="line">            <span class="built_in">panic</span>(err)</span><br><span class="line">        }</span><br><span class="line">        writer.Write(all)</span><br><span class="line">    })</span><br><span class="line"></span><br><span class="line">    err := http.ListenAndServe(<span class="string">":8080"</span>, <span class="literal">nil</span>)</span><br><span class="line">    <span class="keyword">if</span> err != <span class="literal">nil</span> {</span><br><span class="line">        <span class="built_in">panic</span>(err)</span><br><span class="line">    }</span><br><span class="line">}</span><br></pre></td></tr></tbody></table></figure>

<figure class="highlight shell"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">&gt;</span><span class="bash">curl http://localhost:8080/list/fib.txt</span></span><br><span class="line">1</span><br><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">5</span><br><span class="line">8</span><br><span class="line">13</span><br><span class="line">21</span><br><span class="line">34</span><br><span class="line">55</span><br><span class="line">89</span><br><span class="line">144</span><br><span class="line">233</span><br><span class="line">377</span><br><span class="line">610</span><br><span class="line">987</span><br><span class="line">1597</span><br><span class="line">2584</span><br><span class="line">4181</span><br><span class="line">6765</span><br><span class="line"></span><br><span class="line"><span class="meta">&gt;</span><span class="bash">curl http://localhost:8080/list/fib2.txt</span></span><br><span class="line">open fib.txta: The system cannot find the file specified.</span><br></pre></td></tr></tbody></table></figure>

<figure class="highlight go"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">// 业务逻辑</span></span><br><span class="line"><span class="keyword">package</span> server</span><br><span class="line"></span><br><span class="line"><span class="keyword">import</span> (</span><br><span class="line">    <span class="string">"io/ioutil"</span></span><br><span class="line">    <span class="string">"net/http"</span></span><br><span class="line">    <span class="string">"os"</span></span><br><span class="line">)</span><br><span class="line"></span><br><span class="line"><span class="comment">// 把业务逻辑分离开来，如果有错误，就返回一个错误，</span></span><br><span class="line"><span class="function"><span class="keyword">func</span> <span class="title">HandleFileList</span><span class="params">(writer http.ResponseWriter, request *http.Request)</span> <span class="title">error</span></span> {</span><br><span class="line">    path := request.URL.Path[<span class="built_in">len</span>(<span class="string">"/list/"</span>):]</span><br><span class="line">    file, err := os.Open(path)</span><br><span class="line">    <span class="keyword">if</span> err != <span class="literal">nil</span> {</span><br><span class="line">        <span class="comment">// panic err</span></span><br><span class="line">        <span class="keyword">return</span> err</span><br><span class="line">    }</span><br><span class="line">    <span class="keyword">defer</span> file.Close()</span><br><span class="line">    all, err := ioutil.ReadAll(file)</span><br><span class="line">    <span class="keyword">if</span> err != <span class="literal">nil</span> {</span><br><span class="line">        <span class="keyword">return</span> err</span><br><span class="line">    }</span><br><span class="line">    writer.Write(all)</span><br><span class="line">    <span class="keyword">return</span> <span class="literal">nil</span> <span class="comment">// 没有错误返回nil</span></span><br><span class="line">}</span><br></pre></td></tr></tbody></table></figure>

<figure class="highlight go"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">package</span> main</span><br><span class="line"></span><br><span class="line"><span class="keyword">import</span> (</span><br><span class="line">    <span class="string">"io/ioutil"</span></span><br><span class="line">    <span class="string">"learn2/learnerror/server"</span></span><br><span class="line">    <span class="string">"net/http"</span></span><br><span class="line">    <span class="string">"os"</span></span><br><span class="line">)</span><br><span class="line"></span><br><span class="line"><span class="comment">//  定义一下这个函数类型</span></span><br><span class="line"><span class="keyword">type</span> appHandle <span class="function"><span class="keyword">func</span><span class="params">(writer http.ResponseWriter, request *http.Request)</span> <span class="title">error</span></span></span><br><span class="line"></span><br><span class="line"><span class="comment">// 错误统一处理 函数式编程，入参是一个函数，出参也是一个函数</span></span><br><span class="line"><span class="function"><span class="keyword">func</span> <span class="title">errorWrapper</span><span class="params">(handle appHandle)</span> <span class="title">func</span><span class="params">(http.ResponseWriter, *http.Request)</span></span> {</span><br><span class="line">    <span class="keyword">return</span> <span class="function"><span class="keyword">func</span><span class="params">(writer http.ResponseWriter, request *http.Request)</span></span> {</span><br><span class="line">        err := handle(writer, request)</span><br><span class="line">        <span class="keyword">if</span> err != <span class="literal">nil</span> { <span class="comment">// 错误统一处理</span></span><br><span class="line">            <span class="comment">// system error</span></span><br><span class="line">            code := http.StatusOK</span><br><span class="line">            <span class="keyword">switch</span> {</span><br><span class="line">            <span class="keyword">case</span> os.IsNotExist(err):</span><br><span class="line">                code = http.StatusNotFound</span><br><span class="line">            <span class="keyword">default</span>:</span><br><span class="line">                code = http.StatusInternalServerError</span><br><span class="line">            }</span><br><span class="line">            http.Error(writer, http.StatusText(code), code) <span class="comment">// 第二个参数不暴露内部接口</span></span><br><span class="line">        }</span><br><span class="line">    }</span><br><span class="line">}</span><br><span class="line"></span><br><span class="line"><span class="function"><span class="keyword">func</span> <span class="title">main</span><span class="params">()</span></span> {</span><br><span class="line">    <span class="comment">// 用errorWrapper包装一下这个业务逻辑函数</span></span><br><span class="line">    http.HandleFunc(<span class="string">"/list/"</span>, errorWrapper(server.HandleFileList))</span><br><span class="line">    err := http.ListenAndServe(<span class="string">":8080"</span>, <span class="literal">nil</span>)</span><br><span class="line">    <span class="keyword">if</span> err != <span class="literal">nil</span> {</span><br><span class="line">        <span class="built_in">panic</span>(err)</span><br><span class="line">    }</span><br><span class="line">}</span><br></pre></td></tr></tbody></table></figure>

<figure class="highlight shell"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">&gt;</span><span class="bash">curl http://localhost:8080/list/fib.txt</span></span><br><span class="line">1</span><br><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">5</span><br><span class="line">8</span><br><span class="line">13</span><br><span class="line">21</span><br><span class="line">34</span><br><span class="line">55</span><br><span class="line">89</span><br><span class="line">144</span><br><span class="line">233</span><br><span class="line">377</span><br><span class="line">610</span><br><span class="line">987</span><br><span class="line">1597</span><br><span class="line">2584</span><br><span class="line">4181</span><br><span class="line">6765</span><br></pre></td></tr></tbody></table></figure>

<hr>
<figure class="highlight go"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">package</span> filelisting</span><br><span class="line"></span><br><span class="line"><span class="keyword">import</span> (</span><br><span class="line">    <span class="string">"fmt"</span></span><br><span class="line">    <span class="string">"io/ioutil"</span></span><br><span class="line">    <span class="string">"net/http"</span></span><br><span class="line">    <span class="string">"os"</span></span><br><span class="line">    <span class="string">"strings"</span></span><br><span class="line">)</span><br><span class="line"></span><br><span class="line"><span class="keyword">const</span> prefix = <span class="string">"/list/"</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">type</span> userError <span class="keyword">string</span></span><br><span class="line"></span><br><span class="line"><span class="function"><span class="keyword">func</span> <span class="params">(e userError)</span> <span class="title">Error</span><span class="params">()</span> <span class="title">string</span></span> {</span><br><span class="line">    <span class="keyword">return</span> e.Message()</span><br><span class="line">}</span><br><span class="line"></span><br><span class="line"><span class="function"><span class="keyword">func</span> <span class="params">(e userError)</span> <span class="title">Message</span><span class="params">()</span> <span class="title">string</span></span> {</span><br><span class="line">    <span class="keyword">return</span> <span class="keyword">string</span>(e)</span><br><span class="line">}</span><br><span class="line"></span><br><span class="line"><span class="comment">// 把业务逻辑分离开来，如果有错误，就返回一个错误</span></span><br><span class="line"><span class="function"><span class="keyword">func</span> <span class="title">HandleFileList</span><span class="params">(writer http.ResponseWriter, request *http.Request)</span> <span class="title">error</span></span> {</span><br><span class="line">    fmt.Println()</span><br><span class="line">    <span class="comment">// 如果不是以list开头，我们自己抛出一个自己定义的代码</span></span><br><span class="line">    <span class="keyword">if</span> strings.Index(request.URL.Path, prefix) != <span class="number">0</span> {</span><br><span class="line">        <span class="comment">//return errors.New(request.URL.Path + "path %s must start " + "with %s" + prefix)</span></span><br><span class="line">        <span class="keyword">return</span> userError(fmt.Sprintf(<span class="string">"path %s must start "</span>+<span class="string">"with %s"</span>,</span><br><span class="line">            request.URL.Path, prefix))</span><br><span class="line">        <span class="comment">//path /abc must start with /list/</span></span><br><span class="line">    }</span><br><span class="line">    path := request.URL.Path[<span class="built_in">len</span>(prefix):] <span class="comment">// 这里可能会切片下标越界</span></span><br><span class="line">    file, err := os.Open(path)</span><br><span class="line">    <span class="keyword">if</span> err != <span class="literal">nil</span> {</span><br><span class="line">        <span class="comment">//panic err</span></span><br><span class="line">        <span class="keyword">return</span> err</span><br><span class="line">    }</span><br><span class="line">    <span class="keyword">defer</span> file.Close()</span><br><span class="line">    all, err := ioutil.ReadAll(file)</span><br><span class="line">    <span class="keyword">if</span> err != <span class="literal">nil</span> {</span><br><span class="line">        <span class="keyword">return</span> err</span><br><span class="line">    }</span><br><span class="line"></span><br><span class="line">    writer.Write(all)</span><br><span class="line">    <span class="keyword">return</span> <span class="literal">nil</span> <span class="comment">// 没有错误返回nil</span></span><br><span class="line">}</span><br></pre></td></tr></tbody></table></figure>

<figure class="highlight go"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">package</span> main</span><br><span class="line"></span><br><span class="line"><span class="keyword">import</span> (</span><br><span class="line">    <span class="string">"log"</span></span><br><span class="line">    <span class="string">"net/http"</span></span><br><span class="line">    _ <span class="string">"net/http/pprof"</span></span><br><span class="line">    <span class="string">"os"</span></span><br><span class="line"></span><br><span class="line">    <span class="string">"imooc.com/ccmouse/learngo/lang/errhandling/filelistingserver/filelisting"</span></span><br><span class="line">)</span><br><span class="line"></span><br><span class="line"><span class="comment">// 接口 用户可以看到的error类型</span></span><br><span class="line"><span class="keyword">type</span> userError <span class="keyword">interface</span> {</span><br><span class="line">    error            <span class="comment">// 系统的错误</span></span><br><span class="line">    Message() <span class="keyword">string</span> <span class="comment">// 定义用户可以看的错误</span></span><br><span class="line">}</span><br><span class="line"></span><br><span class="line"><span class="comment">//  定义一下这个函数类型</span></span><br><span class="line"><span class="keyword">type</span> appHandler <span class="function"><span class="keyword">func</span><span class="params">(writer http.ResponseWriter, request *http.Request)</span> <span class="title">error</span></span></span><br><span class="line"></span><br><span class="line"><span class="comment">// 错误统一处理 函数式编程，入参是一个函数，出参也是一个函数</span></span><br><span class="line"><span class="function"><span class="keyword">func</span> <span class="title">errWrapper</span><span class="params">(handler appHandler)</span> <span class="title">func</span><span class="params">(http.ResponseWriter, *http.Request)</span></span> {</span><br><span class="line">    <span class="keyword">return</span> <span class="function"><span class="keyword">func</span><span class="params">(writer http.ResponseWriter, request *http.Request)</span></span> {</span><br><span class="line">        <span class="comment">// 虽然说http里面发生错误会自己recover 但是我们还是自己做下处理比较好</span></span><br><span class="line">        <span class="keyword">defer</span> <span class="function"><span class="keyword">func</span><span class="params">()</span></span> {</span><br><span class="line">            <span class="keyword">if</span> r := <span class="built_in">recover</span>(); r != <span class="literal">nil</span> {</span><br><span class="line">                log.Printf(<span class="string">"Panic: %v"</span>, r) <span class="comment">// 在ide终端和浏览器输出</span></span><br><span class="line">                http.Error(writer,</span><br><span class="line">                    http.StatusText(http.StatusInternalServerError),</span><br><span class="line">                    http.StatusInternalServerError)</span><br><span class="line">            }</span><br><span class="line">        }()</span><br><span class="line"></span><br><span class="line">        err := handler(writer, request)</span><br><span class="line">        <span class="keyword">if</span> err != <span class="literal">nil</span> { <span class="comment">// 错误统一处理</span></span><br><span class="line">            log.Printf(<span class="string">"Error occurred "</span>+<span class="string">"handling request: %s %v"</span>, err.Error(), err) <span class="comment">// 在ide终端输出</span></span><br><span class="line"></span><br><span class="line">            <span class="comment">// user error  自己定义的错误，有一些要返回给用户，可以这样做</span></span><br><span class="line">            <span class="keyword">if</span> userErr, ok := err.(userError); ok {  <span class="comment">// 判断err是否属于某个接口类型</span></span><br><span class="line">                http.Error(writer, userErr.Message(), http.StatusBadRequest)</span><br><span class="line">                <span class="keyword">return</span></span><br><span class="line">            }</span><br><span class="line"></span><br><span class="line">            <span class="comment">// system error</span></span><br><span class="line">            code := http.StatusOK</span><br><span class="line">            <span class="keyword">switch</span> {</span><br><span class="line">            <span class="keyword">case</span> os.IsNotExist(err):</span><br><span class="line">                code = http.StatusNotFound</span><br><span class="line">            <span class="keyword">case</span> os.IsPermission(err):</span><br><span class="line">                code = http.StatusForbidden</span><br><span class="line">            <span class="keyword">default</span>:</span><br><span class="line">                code = http.StatusInternalServerError</span><br><span class="line">            }</span><br><span class="line">            http.Error(writer, http.StatusText(code), code) <span class="comment">// 第二个参数不暴露内部接口</span></span><br><span class="line">        }</span><br><span class="line">    }</span><br><span class="line">}</span><br><span class="line"></span><br><span class="line"><span class="function"><span class="keyword">func</span> <span class="title">main</span><span class="params">()</span></span> {</span><br><span class="line">    <span class="comment">// 用errorWrapper包装一下这个业务逻辑函数</span></span><br><span class="line">    <span class="comment">// 让 / 下的网页都可以访问，</span></span><br><span class="line">    http.HandleFunc(<span class="string">"/"</span>, errWrapper(filelisting.HandleFileList))</span><br><span class="line">    err := http.ListenAndServe(<span class="string">":8080"</span>, <span class="literal">nil</span>)</span><br><span class="line">    <span class="keyword">if</span> err != <span class="literal">nil</span> {</span><br><span class="line">        <span class="built_in">panic</span>(err)</span><br><span class="line">    }</span><br><span class="line">}</span><br></pre></td></tr></tbody></table></figure>

<figure class="highlight go"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="string">``</span><span class="string">`</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">`</span><span class="string">``</span><span class="keyword">go</span></span><br></pre></td></tr></tbody></table></figure>

<figure class="highlight go"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br></pre></td><td class="code"><pre><span class="line"><span class="string">``</span><span class="string">`</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">### 5. 错误逐层传递时，层层都加日志</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">根据笔者经验，层层都加日志非常方便故障定位。</span></span><br><span class="line"><span class="string">说明：至于通过测试来发现故障，而不是日志，目前很多团队还很难做到。如果你或你的团队能做到，那么请忽略这个姿势:)</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">### 6. 错误处理使用defer</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">我们一般通过判断error的值来处理错误，如果当前操作失败，需要将本函数中已经create的资源destroy掉，示例代码如下：</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">`</span><span class="string">``</span><span class="keyword">go</span></span><br><span class="line"><span class="function"><span class="keyword">func</span> <span class="title">deferDemo</span><span class="params">()</span> <span class="title">error</span></span> {</span><br><span class="line">    err := createResource1()</span><br><span class="line">    <span class="keyword">if</span> err != <span class="literal">nil</span> {</span><br><span class="line">        <span class="keyword">return</span> ERR_CREATE_RESOURCE1_FAILED</span><br><span class="line">    }</span><br><span class="line">    err = createResource2()</span><br><span class="line">    <span class="keyword">if</span> err != <span class="literal">nil</span> {</span><br><span class="line">        destroyResource1()</span><br><span class="line">        <span class="keyword">return</span> ERR_CREATE_RESOURCE2_FAILED</span><br><span class="line">    }</span><br><span class="line">    err = createResource3()</span><br><span class="line">    <span class="keyword">if</span> err != <span class="literal">nil</span> {</span><br><span class="line">        destroyResource1()</span><br><span class="line">        destroyResource2()</span><br><span class="line">        <span class="keyword">return</span> ERR_CREATE_RESOURCE3_FAILED</span><br><span class="line">    }</span><br><span class="line">    err = createResource4()</span><br><span class="line">    <span class="keyword">if</span> err != <span class="literal">nil</span> {</span><br><span class="line">        destroyResource1()</span><br><span class="line">        destroyResource2()</span><br><span class="line">        destroyResource3()</span><br><span class="line">        <span class="keyword">return</span> ERR_CREATE_RESOURCE4_FAILED</span><br><span class="line">    }</span><br><span class="line">    <span class="keyword">return</span> <span class="literal">nil</span></span><br><span class="line">}</span><br></pre></td></tr></tbody></table></figure>

<p>当Golang的代码执行时，如果遇到defer的闭包调用，则压入堆栈。当函数返回时，会按照后进先出的顺序调用闭包。<br>对于闭包的参数是值传递，而对于外部变量却是引用传递，所以闭包中的外部变量err的值就变成外部函数返回时最新的err值。<br>根据这个结论，我们重构上面的示例代码：</p>
<figure class="highlight go"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="keyword">func</span> <span class="title">deferDemo</span><span class="params">()</span> <span class="title">error</span></span> {</span><br><span class="line">    err := createResource1()</span><br><span class="line">    <span class="keyword">if</span> err != <span class="literal">nil</span> {</span><br><span class="line">        <span class="keyword">return</span> ERR_CREATE_RESOURCE1_FAILED</span><br><span class="line">    }</span><br><span class="line">    <span class="keyword">defer</span> <span class="function"><span class="keyword">func</span><span class="params">()</span></span> {</span><br><span class="line">        <span class="keyword">if</span> err != <span class="literal">nil</span> {</span><br><span class="line">            destroyResource1()</span><br><span class="line">        }</span><br><span class="line">    }()</span><br><span class="line">    err = createResource2()</span><br><span class="line">    <span class="keyword">if</span> err != <span class="literal">nil</span> {</span><br><span class="line">        <span class="keyword">return</span> ERR_CREATE_RESOURCE2_FAILED</span><br><span class="line">    }</span><br><span class="line">    <span class="keyword">defer</span> <span class="function"><span class="keyword">func</span><span class="params">()</span></span> {</span><br><span class="line">        <span class="keyword">if</span> err != <span class="literal">nil</span> {</span><br><span class="line">            destroyResource2()</span><br><span class="line">        }</span><br><span class="line">    }()</span><br><span class="line">    err = createResource3()</span><br><span class="line">    <span class="keyword">if</span> err != <span class="literal">nil</span> {</span><br><span class="line">        <span class="keyword">return</span> ERR_CREATE_RESOURCE3_FAILED</span><br><span class="line">    }</span><br><span class="line">    <span class="keyword">defer</span> <span class="function"><span class="keyword">func</span><span class="params">()</span></span> {</span><br><span class="line">        <span class="keyword">if</span> err != <span class="literal">nil</span> {</span><br><span class="line">            destroyResource3()</span><br><span class="line">        }</span><br><span class="line">    }()</span><br><span class="line">    err = createResource4()</span><br><span class="line">    <span class="keyword">if</span> err != <span class="literal">nil</span> {</span><br><span class="line">        <span class="keyword">return</span> ERR_CREATE_RESOURCE4_FAILED</span><br><span class="line">    }</span><br><span class="line">    <span class="keyword">return</span> <span class="literal">nil</span></span><br><span class="line">}</span><br></pre></td></tr></tbody></table></figure>

<h3 id="7-当发生错误时，不忽略有用的返回值"><a href="#7-当发生错误时，不忽略有用的返回值" class="headerlink" title="7. 当发生错误时，不忽略有用的返回值"></a>7. 当发生错误时，不忽略有用的返回值</h3><p>通常，当函数返回non-nil的error时，其他的返回值是未定义的(undefined)，这些未定义的返回值应该被忽略。然而，有少部分函数在发生错误时，仍然会返回一些有用的返回值。比如，当读取文件发生错误时，Read函数会返回可以读取的字节数以及错误信息。对于这种情况，应该将读取到的字符串和错误信息一起打印出来。</p>
<figure class="highlight go"><table><tbody><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"></span><br></pre></td></tr></tbody></table></figure>

<h2 id="Golang错误和异常是可以互相转换的："><a href="#Golang错误和异常是可以互相转换的：" class="headerlink" title="Golang错误和异常是可以互相转换的："></a>Golang错误和异常是可以互相转换的：</h2><ol>
<li>错误转异常，比如程序逻辑上尝试请求某个URL，最多尝试三次，尝试三次的过程中请求失败是错误，尝试完第三次还不成功的话，失败就被提升为异常了。</li>
<li>异常转错误，比如panic触发的异常被recover恢复后，将返回值中error类型的变量进行赋值，以便上层函数继续走错误处理流程。</li>
</ol>
<script>
        document.querySelectorAll('.github-emoji')
          .forEach(el => {
            if (!el.dataset.src) { return; }
            const img = document.createElement('img');
            img.style = 'display:none !important;';
            img.src = el.dataset.src;
            img.addEventListener('error', () => {
              img.remove();
              el.style.color = 'inherit';
              el.style.backgroundImage = 'none';
              el.style.background = 'none';
            });
            img.addEventListener('load', () => {
              img.remove();
            });
            document.body.appendChild(img);
          });
      </script>
    </div>

    
    
    

    <footer class="post-footer">

        

          <div class="post-nav">
            <div class="post-nav-item">
                <a href="/2021/07/05/go-han-shu/" rel="prev" title="go函数">
                  <i class="fa fa-chevron-left"></i> go函数
                </a>
            </div>
            <div class="post-nav-item">
                <a href="/2021/07/06/go-ni-ming-han-shu/" rel="next" title="go匿名函数">
                  go匿名函数 <i class="fa fa-chevron-right"></i>
                </a>
            </div>
          </div>
    </footer>
  </article>
</div>






</div>
  </main>

  <footer class="footer">
    <div class="footer-inner">


<div class="copyright">
  &copy; 
  <span itemprop="copyrightYear">2021</span>
  <span class="with-love">
    <i class="fa fa-heart"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">外心人D</span>
</div>
  <div class="powered-by">由 <a href="https://hexo.io/" rel="noopener" target="_blank">Hexo</a> & <a href="https://theme-next.js.org/muse/" rel="noopener" target="_blank">NexT.Muse</a> 强力驱动
  </div>

    </div>
  </footer>

  
  <script src="https://cdn.jsdelivr.net/npm/animejs@3.2.1/lib/anime.min.js" integrity="sha256-XL2inqUJaslATFnHdJOi9GfQ60on8Wx1C2H8DYiN1xY=" crossorigin="anonymous"></script>
<script src="/js/comments.js"></script><script src="/js/utils.js"></script><script src="/js/motion.js"></script><script src="/js/schemes/muse.js"></script><script src="/js/next-boot.js"></script>

  




  





</body>
</html>
